### DISCUSSION TITLE
Uncaught Error : Facing an issue in React Native

### DISCUSSION THREAD

**Discussion by afkcodes** [pmndrs/valtio#764] [Q&A]
Hello Everyone I have been using this amazing lib for the first time , i have tried to follow the docs and the setup seems as per the doc, but whenever i am trying to update the state, its saying the following error : 

```TypeError: target property is non-configurable and non-writable, and set trap value differs from target property value, js engine: hermes```

Has anyone faced this before or is this a issue that needs to be reported

This is how i have constituted my state
![Screenshot 2023-07-21 at 12 48 45 AM](https://github.com/pmndrs/valtio/assets/17390257/e022cf3c-5bd5-4923-998d-a371f2367b95)

and i am updating it like this:
![Screenshot 2023-07-21 at 12 50 00 AM](https://github.com/pmndrs/valtio/assets/17390257/3d46da14-a951-457a-9d33-9224ca280b19)

**Comment by afkcodes** [comment]
Found that v1.10.6 works fine with react native, but 1.10.7 is giving error i will check with 1.11.0 and update here.

**Reply by dai-shi** [reply]
That's great to know. I don't think 1.11.0 changes it. It's probably caused by #752.
Assuming Hermes supports `Object.preventExtensions`, do you add a new property on snapshots somewhere?

**Reply by afkcodes** [reply]
no I'm not but i will recheck the code and let you know, it's giving error on every property that is being updated

**Reply by afkcodes** [reply]
for eg: it gives error when the i call playerActions.setStatus, similarly for others as well.

**Reply by dai-shi** [reply]
Would you be able to to create a very small reproduction with Expo Snack?

**Reply by afkcodes** [reply]
should i share my repo with you. ? there's nothing at the moment except the valtio addition

**Reply by afkcodes** [reply]
snap is kind of behaving wierd, i guess nothing breaks there, even if i do it wrong knowingly its working there.
here is the snap: where i am adding property on the go to the state: https://snack.expo.dev/@ashishfeels/6a3c8a (incorrect i guess)

**Reply by afkcodes** [reply]
looked into my code, i think there is nothing that is messing up the things, to throw this error

**Reply by dai-shi** [reply]
> Assuming Hermes supports `Object.preventExtensions`

The assumption wasn't correct, according to: https://github.com/pmndrs/valtio/pull/752#issuecomment-1645125261
cc @lovetingyuan @pastelmind

**Reply by afkcodes** [reply]
Thanks for the update @dai-shi , so are we going to revert the commit or we are going to dig this further, should i downgrade to 1.10.6 for the moment.

**Reply by dai-shi** [reply]
I think we should revert #752 for now. 
Can @pastelmind open a PR?

**Reply by pastelmind** [reply]
On second thought, I would like to investigate this a bit further.

> snap is kind of behaving wierd, i guess nothing breaks there, even if i do it wrong knowingly its working there. here is the snap: where i am adding property on the go to the state: https://snack.expo.dev/@ashishfeels/6a3c8a (incorrect i guess)

@afkcodes I opened your Expo Snack example and tried it on iOS/Android/Web (emulated) as well as on my own Android phone (using Expo Go). In all cases, the app executed well without error.

I even tried configuring the JS engine to Hermes in the snack by adding `app.json`:

```json
{
  "expo": {
    "jsEngine": "hermes"
  }
}
```

...but the example still executed just fine.

Can you provide detailed instructions on reproducing the error? Please understand that I have never used Expo or React Native before.

Edit: Nevermind, I reproduced it by running a fresh project created with `create-expo-app`, then adding Valtio. Will report back after further investigation.

**Reply by pastelmind** [reply]
Here is my reproduction.

<details>
<summary>Click to see my code</summary>

```jsx
// App.js
import { StatusBar } from "expo-status-bar";
import { StyleSheet, Text, View } from "react-native";
import { proxy, useSnapshot } from "valtio";

const store = proxy({ count: 0, untouched: "asdf" });

const incrementCounter = () => store.count++;

const reportStatus = (value) => ({
  extensible: Object.isExtensible(value),
  sealed: Object.isSealed(value),
  frozen: Object.isFrozen(value),
});

export default function App() {
  const snap = useSnapshot(store);

  console.log("count:", snap.count);
  console.log("store descriptors:", Object.getOwnPropertyDescriptors(store));
  console.log("store status:", reportStatus(store));
  console.log("snapshot descriptors:", Object.getOwnPropertyDescriptors(snap));
  console.log("snapshot status:", reportStatus(snap));

  return (
    <View style={styles.container}>
      <Text onPress={incrementCounter}>MyCount: {snap.count}</Text>
      <StatusBar style="auto" />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
});
```
</details>

When running the code with Valtio 1.11.0, I noticed that the `count` property is being made non-configurable, non-enumerable, and non-writable (i.e. non-CEW) after the first mutation. Properties that are not mutated are still CEW.

<details>
<summary>See log</summary>

```
 LOG  count: 0
 LOG  store descriptors: {"count": {"configurable": true, "enumerable": true, "value": 0, "writable": true}, "untouched": {"configurable": true, "enumerable": true, "value": "asdf", "writable": true}}
 LOG  store status: {"extensible": true, "frozen": false, "sealed": false}
 LOG  snapshot descriptors: {"count": {"configurable": true, "enumerable": true, "value": 0, "writable": false}, "untouched": {"configurable": true, "enumerable": true, "value": "asdf", "writable": false}}
 LOG  snapshot status: {"extensible": false, "frozen": false, "sealed": false}
 LOG  count: 1
 LOG  store descriptors: {"count": {"configurable": false, "enumerable": false, "value": 1, "writable": false}, "untouched": {"configurable": true, "enumerable": true, "value": "asdf", "writable": true}}
 LOG  store status: {"extensible": true, "frozen": false, "sealed": false}
 LOG  snapshot descriptors: {"count": {"configurable": true, "enumerable": true, "value": 1, "writable": false}, "untouched": {"configurable": true, "enumerable": true, "value": "asdf", "writable": false}}
 LOG  snapshot status: {"extensible": false, "frozen": false, "sealed": false}
```
</details>

I reproduced the error using Valtio 1.11.0 but not 1.10.7. Also, given that the bug affects the _proxied store_ and not the _snapshot_, I suspect that #752 (which should only affect snapshots) may not be the cause after all.

@afkcodes Can you double-check if the bug really occurs on 1.10.7?

**Reply by pastelmind** [reply]
I may have nailed the problem down to https://github.com/facebook/hermes/issues/1065.

**Reply by afkcodes** [reply]
yes @pastelmind  i can confirm that same problem occurred on 1.10.7, on snack i was not able to reproduce the same but that is the exact replica of the code i have written it fails on device. at the moment i have downgraded to 1.10.6 and everything seems to be working great

**Reply by pastelmind** [reply]
Hmm. Perhaps the issue I found may be a different one from yours. @afkcodes Can you please point me to a repository so I can clone it locally and examine the issue?

**Reply by afkcodes** [reply]
@pastelmind sure let me make it public, for some time. please let me know when you are done with your testing.

**Reply by afkcodes** [reply]
Here you go : 
https://github.com/highfive-digital/tarana/


There is a branch name called - /feature/state-and-player

check out to this branch and please to a manual install of `yarn add react-native-track-player@nightly` or else it might throw an error

**Reply by pastelmind** [reply]
I tested your app on my own real Android phone in development mode. By clicking on the buttons labelled "1", "5", and "10" (below the "SetSleepTimer Minutes"), I reproduced the following error with Valtio 1.11.0 but NOT Valtio 1.10.7. Are you really really sure that Valtio 1.10.7 is the breaking version?

<img width="300" alt="Screenshot_20230724_003857_Tarana" src="https://github.com/pmndrs/valtio/assets/29533079/04f1b408-f5c2-409a-a936-d5ad9af31717"/>

My setup is:

- Debugging PC
  - Model: MacBook Pro 16-in, 2021 (Apple M1 Pro)
  - OS: macOS Ventura 13.4.1 (c)
  - JVM: Zulu 11 (installed [following the official docs](https://reactnative.dev/docs/environment-setup))
  - Node.js: 18.17.0
- Debugging Phone
  - Model: Galaxy S21 (SM-G991N)
  - OS: Android 13

Mind you that I have never used Android Studio before so I'm unfamiliar with how all this works.

**Reply by afkcodes** [reply]
yeah im sure, don't know why though. if you can see the history i have downgraded from 1.10.7

**Reply by afkcodes** [reply]
there is no issues with your setup. can it be some other issue why i am facing this ? yarn/npm cache. i have also tried deleting yarn cache, lock file, node_moduoes folder and clearing metro cache as well.

**Reply by pastelmind** [reply]
I asked a fellow developer with React Native experience for help. He verified that v1.10.7 is OK but v1.11.0 is bugged.

@afkcodes I'm afraid I don't know why your results are different. Unfortunately my friend could not explain either. We could release a new version that reverts v1.11.0 (#760) and verify that it works for you.

In the meantime, I suggest you try creating a new React Native project on your development environment and check if Valtio 1.10.7 is bugged. You can use [my sample code](https://github.com/pmndrs/valtio/discussions/764#discussioncomment-6517634) to reproduce the bug.

**Reply by lovetingyuan** [reply]
```js
import { proxy } from 'valtio'

const store = proxy({
  foo: 2,
})

store.foo = {}
```

this code throws error(Uncaught TypeError: target property is non-configurable and non-writable, and set trap value differs from target property value) in `1.11.0` but there is no error in `1.10.7`(tested via hermes-cli)

While `store.foo = 3` works well.

remove `defineProperty` hook in `handler`, there is no error.

**Reply by afkcodes** [reply]
i wil check that out @pastelmind  for sure by eod. Thanks

**Reply by afkcodes** [reply]
hey @pastelmind @dai-shi , today when i upgraded back to 1.10.7 it seems to work fine, i am baffled now what was the issue earlier that it was throwing error. Sorry this might be yarn or my PC doing some weird effect. i will check this once again and report but now everything seems to work fine.

**Reply by dai-shi** [reply]
Sorry for the late response, so in summary, #752 is fine and #760 causes an issue with Hermes? (Or, is it not only with Hermes?)

**Reply by pastelmind** [reply]
The issue is specific to Hermes. Chrome (V8), Firefox (Spidermonkey), and Safari (JavaScriptCore) are OK.

**Reply by afkcodes** [reply]
just an update I'm still on 1.10.7 and it works fine.
